Add glob test to exclude

einstine909 opened the pull request:
This implements #245

Let me know if I need to change anything.

themylogin commented on README.md:
```diff
@@ -51,6 +51,7 @@ periodic-snapshot-tasks:
     # independently but in one transaction.
     exclude:
       - data/src/excluded
+	  - data/*/excluded
```
Let's be consistent with tabs and spaces here (and below) please

themylogin commented on tests/dataset/test_exclude.py:
```diff
@@ -11,6 +11,8 @@
     ("my-data", ["data/.system"], False),
     ("my-data/.system", ["data/.system"], False),
+	("data/.system/cores", ["data/*/cores"], True),
+	("data/.system2/cores", ["data/*/cores"], False),
```
Why won't be `data/.system2/cores` excluded by `data/*/cores`?

einstine909 replied:
It should, and the test does indeed fail. I think I meant to put `data/.system/cores2` and ran the test before saving that file.

themylogin commented on zettarepl/dataset/exclude.py:
```diff
@@ -1,5 +1,5 @@
 # -*- coding=utf-8 -*-
-import logging
+import logging, fnmatch
```
Let's follow https://peps.python.org/pep-0008/#imports

einstine909 replied:
Will do.

themylogin reviewed (changes requested):
That's a great idea! I see two areas that need checking:
* How well does this play with `ReplicationTask._validate_exclude` https://github.com/truenas/zettarepl/blob/master/zettarepl/replication/task/task.py#L255 ?
* Does https://github.com/truenas/zettarepl/blob/master/zettarepl/zcp/recursive_snapshot_exclude.lua that is used to create consistent recursive snapshots support this?

einstine909 commented:
> * Does https://github.com/truenas/zettarepl/blob/master/zettarepl/zcp/recursive_snapshot_exclude.lua that is used to create consistent recursive snapshots support this?

It does not, and to my understanding, Lua does not have a built-in POSIX compliant glob function. I do have some ideas:

1. Pass every single excluded dataset found using python's glob function to Lua, kind of similar to how it is done now
2. Include some Lua library that provides a POSIX glob function, like [this one](https://github.com/davidm/lua-glob-pattern)

Thoughts?

themylogin commented:
@einstine909 option 1 sounds great to me!

I would also change the method of passing data to the channel program from argv to reading stdin (or, if that does not work with `zfs program`, a temporary file) because the number of excluded datasets can be pretty large.

einstine909 commented:
I went with a temporary file. From my poking around how the lua program is called by zfs, looks to me that it cannot get access to stdin.

How does one go about running the integration tests on TrueNAS? I have a new VM spun up with a `data` pool. But pytest isn't on the system and easy_install or pip are not installed on the OS either... any advise would be appreciated.

themylogin commented:
@einstine909 use TrueNAS SCALE, it has `install-dev-tools` command which will will install pip

einstine909 commented:
I just pushed my changes to my repo. It turns out that the lua interpreter that zfs uses has no ability to do any IO except modify datasets, with the exception of reading in the command line arguments that it was called with. I now generate a zcp program per snapshot-task that has the excluded datasets hard-coded into the lua program. I added in some additional functionality to help create and send the zcp program over a shell.

I have updated the tests to account for my changes. I found that the test `test__parse_snapshots_name__with_timestamp()` failed on a fresh install of TrueNAS Scale with the current commit [(bfa68c6)](https://github.com/truenas/zettarepl/commit/bfa68c6185db5ea4001beb5800fd15487f12523f) of this repo, so all tests (unit and integration) tests pass, save that one.

themylogin commented:
@einstine909 thank you for your contribution! Sorry it took me so long to review it 😃
